messaging/coap/observe.c

```diff
@@ -1262,47 +1271,104 @@ notify_batch_observer(coap_observer_t *obs, oc_response_t *response)
+#ifdef OC_HAS_FEATURE_ETAG
+  uint64_t etag =
+    oc_ri_get_batch_etag(obs->resource, &obs->endpoint, obs->endpoint.device);
+  if (etag != OC_ETAG_UNINITIALIZED) {
+    memcpy(&response_buffer->etag.value[0], &etag, sizeof(etag));
+    response_buffer->etag.length = sizeof(etag);
+  }
+#endif /* OC_HAS_FEATURE_ETAG */
+  return true;
+}
+
+coap_batch_observer_t *
+coap_get_discovery_batch_observers(void)
 {
-  batch_observer_t *batch_obs =
-    (batch_observer_t *)oc_list_head(g_batch_observers_list);
+  return (coap_batch_observer_t *)oc_list_head(g_batch_observers_list);
+}
+
+void
+coap_process_discovery_batch_observers(void)
+{
+  coap_batch_observer_t *batch_obs =
+    (coap_batch_observer_t *)oc_list_head(g_batch_observers_list);
   if (batch_obs == NULL) {
     return;
   }
-#ifndef OC_DYNAMIC_ALLOCATION
-  uint8_t buffer[OC_MIN_OBSERVE_SIZE];
-#else  /* !OC_DYNAMIC_ALLOCATION */
+#ifdef OC_DYNAMIC_ALLOCATION
   uint8_t *buffer = malloc(OC_MIN_OBSERVE_SIZE);
   if (buffer == NULL) {
     COAP_WRN("out of memory allocating buffer");
     return;
   }
+#else  /* !OC_DYNAMIC_ALLOCATION */
+  uint8_t buffer[OC_MIN_OBSERVE_SIZE];
 #endif /* OC_DYNAMIC_ALLOCATION */
   oc_response_buffer_t response_buffer;
   memset(&response_buffer, 0, sizeof(response_buffer));
   COAP_DBG("Issue GET request to discovery resource for %s resource",
            batch_observer_get_resource_uri(batch_obs));
   response_buffer.buffer = buffer;
   response_buffer.buffer_size = OC_MIN_OBSERVE_SIZE;
+
+  // TODO: cache of etags for batch interface to avoid repeated calculation?
+  // struct {
+  //   const oc_resource_t* resource;
+  //   const oc_endpoint_t* endpoint;
+  //   uint64_t etag;
+  //}
+  // oc_list_t etags;
+
   while (batch_obs != NULL) {
-    if (!batch_obs->resource &&
-        !oc_string_len(batch_obs->removed_resource_uri)) {
-      COAP_WRN("resource is NULL and removed_resource_uri is empty");
-      oc_list_remove(g_batch_observers_list, batch_obs);
-      free_batch_observer(batch_obs);
-      batch_obs = (batch_observer_t *)oc_list_head(g_batch_observers_list);
-      continue;
-    }
-    coap_observer_t *obs = batch_obs->obs;
 #ifdef OC_BLOCK_WISE
+    coap_observer_t *obs = batch_obs->obs;
+    // obs->iface_mask is always OC_IF_B
     oc_string_view_t query = oc_query_encode_interface(obs->iface_mask);
+    // obs->resource is always the discovery resource
     const oc_blockwise_state_t *response_state =
       oc_blockwise_find_response_buffer(oc_string(obs->resource->uri) + 1,
                                         oc_string_len(obs->resource->uri) - 1,
                                         &obs->endpoint, OC_GET, query.data,
                                         query.length, OC_BLOCKWISE_SERVER);
     if (response_state != NULL) {
+      COAP_DBG(
+        "response_state is not NULL, sending of batch response currently "
+        "in progress for endpoint, skipping to next observer");
+      // TODO: reschedule?
```

Comment:
By [RFC ](https://datatracker.ietf.org/doc/html/rfc7959#section-2.4) the current response need to be dropped and new notification need to be send. But with this code the reschedule is planned because it is not removed from the list (line 1392) - we are safe :)

Reply:
Yes, but if nothing happens the notification might be stuck in the list for a long time, I was thinking maybe about scheduling a dispatch in a minute or something.

Reply:
When blockwise transfer finishes then rescheduling is expected via cbk in blockwise.